firedrake/slate/slac/kernel_builder.py

```diff
@@ -199,35 +202,23 @@ def generate_expr_data(expr, temps=None, aux_exprs=None):
     access to all temporaries associated with a particular slate expression.
 
     :arg expression: a :class:`slate.TensorBase` object.
-    :arg temps: a dictionary that becomes populated recursively and is later
-                returned as the temporaries map. This argument is initialized
-                as an empty `dict` before recursion starts.
-    :arg aux_exprs: a list that becomes populated recursively and is later
-                    returned as the list of auxiliary expressions that require
-                    special handling in Slate's linear algebra compiler
-
-    Returns: the arguments temps and aux_exprs.
+
+    Returns: the terminal temporaries map and auxiliary temporaries.
     """
     # Prepare temporaries map and auxiliary expressions list
-    if temps is None:
-        temps = {}
-
-    if aux_exprs is None:
-        aux_exprs = []
-
-    if isinstance(expr, Tensor):
-        temps.setdefault(expr, ast.Symbol("T%d" % len(temps)))
-
-    elif isinstance(expr, TensorOp):
-        # If we have an Action instance, store expr in aux_exprs for
-        # special handling in the compiler
-        if isinstance(expr, Action):
-            aux_exprs.append(expr)
-
-        # Send operands through recursively
-        map(lambda x: generate_expr_data(x, temps=temps,
-                                         aux_exprs=aux_exprs), expr.operands)
-    else:
-        raise NotImplementedError("Type %s not supported." % type(expr))
+    # NOTE: Ordering here matters, especially when running
+    # Slate in parallel.
+    temps = OrderedDict()
+    aux_exprs = OrderedDict()
+    for tensor in traverse_dags([expr]):
+        if isinstance(tensor, Tensor):
+            temps.setdefault(tensor, ast.Symbol("T%d" % len(temps)))
+
+        elif isinstance(tensor, TensorOp):
+            # For Action, we need to declare a temporary later on for the
+            # acting coefficient. For inverses, we may declare additional
+            # temporaries (depending on reference count).
+            if isinstance(tensor, (Action, Inverse)):
+                aux_exprs.setdefault(tensor)
```

Comment:
> It's generating an ordered, set-like object of expressions that need extra reasoning.

So the oddity is that here you call `setdefault` without actually giving a default value. A few lines above you do specify a value.

> Actions require creating temporaries for firedrake.Functions

I guess you do this because actees are non-child data of an `Action`. Still it smells a bit like shoving different things into the same box.

> and Inverse may need a temporaries if it's reused multiple times.

Doesn't this rule apply to all expressions?